integration-tests/byoid/integration_test.go

```diff
@@ -0,0 +1,211 @@
+
+// +build integration
+
+// To run this test locally, you will need to do the following:
+// • Navigate to your Google Cloud Project
+// • Get a copy of the Service Account Key File from somebody
+// • If you are unable to obtain an existing key file, create one:
+//    • > IAM and Admin > Service Accounts
+//    • Under the needed Service Account > Actions > Manage Keys
+//    • Add Key > Create New Key
+//    • Select JSON, and the click Create
+// • > Compute > Compute Engine > VM Instances
+// • Look for an available VM Instance, or create one
+// • On the VM Instance, click the SSH Button
+// • Upload your Service Account Key File
+// • Upload this script, along with setup.sh
+// • Get a copy of the needed environment variables from somebody, and upload those too
+// • Set your environment variables (Usually this will be `source env.conf`)
+// • If the setup script has not yet been run, then run it
+// • `go test -tags integration`
+
+package byoid
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"testing"
+
+	"golang.org/x/oauth2"
+	"golang.org/x/oauth2/google"
+	"google.golang.org/api/dns/v1"
+	"google.golang.org/api/iamcredentials/v1"
+)
+
+const (
+	envCredentials  = "GOOGLE_APPLICATION_CREDENTIALS"
+	envAudienceOIDC = "GCLOUD_TESTS_GOLANG_AUDIENCE_OIDC"
+	envProject      = "GCLOUD_TESTS_GOLANG_PROJECT_ID"
+)
+
+var (
+	oidcAudience string
+	oidcToken    string
+	awsToken     string
+	clientID     string
+	projectID    string
+)
+
+// keyFile is a struct to extract the relevant json fields for our ServiceAccount KeyFile
+type keyFile struct {
+	ClientEmail string `json:"client_email"`
+	ClientID    string `json:"client_id"`
+}
+
+// TestMain contains all of the setup code that needs to be run once before any of the tests are run
+func TestMain(m *testing.M) {
+	keyFileName := os.Getenv(envCredentials)
+	if keyFileName == "" {
+		log.Fatalf("Please set %s to your keyfile", envCredentials)
+	}
+
+	projectID = os.Getenv(envProject)
+	if projectID == "" {
+		log.Fatalf("Please set %s to the ID of the project", envProject)
+	}
+
+	oidcAudience = os.Getenv(envAudienceOIDC)
+	if oidcAudience == "" {
+		log.Fatalf("Please set %s to the OIDC Audience", envAudienceOIDC)
+	}
+
+	kf, err := os.Open(keyFileName)
+	if err != nil {
+		log.Fatalf("Failed to open file '%s': %v", keyFileName, err)
+	}
+	defer kf.Close()
+
+	decoder := json.NewDecoder(kf)
+	var keyFileSettings keyFile
+	err = decoder.Decode(&keyFileSettings)
+	if err != nil {
+		log.Fatalf("Keyfile '%s' stored in improper format: %v", keyFileName, err)
+	}
+
+	clientID = fmt.Sprintf("projects/-/serviceAccounts/%s", keyFileSettings.ClientEmail)
+
+	iam, err := iamcredentials.NewService(context.Background())
+	if err != nil {
+		log.Fatalf("Unable to access iamcredentials service: %v", err)
+	}
+
+	idTokenInfo, err := iam.Projects.ServiceAccounts.GenerateIdToken(clientID, &iamcredentials.GenerateIdTokenRequest{
```

Comment:
If you are not creating an impersonated idtoken you may find [idtoken.NewTokenSource](https://pkg.go.dev/google.golang.org/api/idtoken#NewTokenSource) easier.

Reply:
done